param.go

```diff
@@ -98,9 +98,36 @@ type paramVisitor interface {
 // recursed into.
 type paramVisitorFunc func(param) (recurse bool)
 
-func (f paramVisitorFunc) Visit(p param) paramVisitor {
-	if f(p) {
-		return f
+// paramVisitOnce encapsulates a paramVisitorFunc and a map to track visited nodes.
+type paramVisitOnce struct {
+	f       paramVisitorFunc
+	visited map[string]struct{}
+}
+
+// NewParamVisitOnce constructs a paramVisitor avoiding redundant visits.
+func newParamVisitOnce(visited map[string]struct{}, f paramVisitorFunc) paramVisitor {
```

Comment:
NBD since this is all internal but strictly speaking, newParamVisitOnce
doesn't need to operate on paramVisitorFunc. It can be on any paramVisitor,
and it can track the list of visited nodes internally.

```
type paramVisitorOnce struct {
    visited map[key]struct{}
    visitor paramVisitor
}

func newParamVisitorOnce(paramVisitor) paramVisitor

func (v paramVisitorOnce) Visit(p param) paramVisitor {
    // If visited, return nil
    // else,
    
    if newV := v.visitor.Visit(p); newV != nil {
        v.visitor = newV
        return v
    }
    
    return nil
}
```

This would mean that detectCycles has to use a regular `paramVisitor` instead
of paramVisitorFunc. Or we can change `paramVisitorFunc` to the following and
it'll work with newParamVisitorOnce seamlessly.

    type paramVisitorFunc func(param) paramVisitor

---

That said, this is largely optional since this is all super internal.

Reply:
I actually went down this route originally! However, because `detectCycles` calls `paramVisitorFunc`/`newParamVisitOnce` itself we would lose some of our efficiency due to a new `visited` map getting created each time it recurses. My solution therefore evolved to have the map get passed in. That said, maybe an equally valid solution is to construct the `paramVisitOnce` a single time and pass *that* into `detectCycles`. It would still require the extra parameter to `detectCycles` like this PR adds, but it might clarify some of the internal semantics. Do you think that is preferable?

Reply:
I may be misunderstanding this but under the snippet above, it wouldn't be a
new `visited` map.

In the following code,

```
if newV := v.visitor.Visit(p); newV != nil {
  v.visitor = newV
  return v
}
```

We create a new paramVisitor out of the paramVisitorOnce (that's one alloc to
convert the value into an interface) but re-uses the existing `visited` map.